## NumPy 2

In [1]:
#!pip install numpy

In [2]:
# Import numpy
import numpy as np

In [3]:
# Create numpy arrays from lists
x = np.array([1,2,3])
a = np.array([[1,2,3]])


y = np.array([[3,4,5]])
z = np.array([[6,7],[8,9]])

# Let's take a look at their shapes.
# When working with numpy arrays, .shape will be a very useful debugging tool
print(x.shape)
print(y.shape)
print()
print(z)
print(z.shape)

(3,)
(1, 3)

[[6 7]
 [8 9]]
(2, 2)


Vectors can be represented as 1-D arrays of shape (N,) or 2-D arrays of shape (N, 1) or (1, N). But it's important to note that the shapes (N,), (N, 1), and (1,N) are not the same and may result in different behavior (we'll see some examples below involving matrix multiplication and broadcasting).

Matrices are generally represented as 2-D arrays of shape (M, N).

The best way to ensure your code gives you the behavior you expect is to keep track of your array shapes and try out small test cases or refer back to documentation when you are unsure.

In [4]:
a = np.arange(10)
b = a.reshape((5, 2))
print(a)
print()
print(b)

[0 1 2 3 4 5 6 7 8 9]

[[0 1]
 [2 3]
 [4 5]
 [6 7]
 [8 9]]


### Array Operations

There are many NumPy operations that can be used to reduce a numpy array along an axis.

Let's look at the np.max operation (documentation: https://numpy.org/doc/stable/reference/generated/numpy.ndarray.max.html).

In [5]:
x = np.array([[1,2],[3,4],[5, 6]])
print(x)
print()
print(x.shape)

[[1 2]
 [3 4]
 [5 6]]

(3, 2)


The statement print(np.max(x, axis=1)) involves using the NumPy library. Here's what it does:

- np.max(): This is a function from the NumPy library that calculates the maximum value along a specified axis of an array.
- x: This presumably refers to a NumPy array.
- axis=1: This parameter specifies the axis along which the maximum value should be calculated. In this case, axis=1 means the maximum value is calculated along the rows of the array.
- print(): This function is used to display the result.

So, print(np.max(x, axis=1)) will calculate the maximum value along each row of the array x and print the resulting array of maximum values.

In [8]:
print(np.max(x, axis = 1)) # axis = 1 because we want to work with the row and not the column

[2 4 6]


In [9]:
print(np.max(x, axis = 1).shape)

(3,)


In [10]:
print(np.max(x, axis = 1, keepdims = True))

[[2]
 [4]
 [6]]


In [11]:
print(np.max(x, axis = 1, keepdims = True).shape)

(3, 1)


Next, let's look at some matrix operations. Let's take an element-wise product (Hadamard product).

In [12]:
A = np.array([[1, 2], [3, 4]])
B = np.array([[3, 3], [3, 3]])
print(A)
print(B)
print("---")
print(A * B)

[[1 2]
 [3 4]]
[[3 3]
 [3 3]]
---
[[ 3  6]
 [ 9 12]]


We can do matrix multiplication with np.matmul or @.

In [31]:
# One way to do matrix multiplication
print(np.matmul(A, B))
print("---")
# Another way to do matrix multiplication
print(A @ B)

[[ 9  9]
 [21 21]]
---
[[ 9  9]
 [21 21]]


We can take the dot product or a matrix vector product with np.dot.

In [32]:
"""
    np.dot(u, v): This is a function call where np is the NumPy module. 
    It explicitly calls the dot() function from the NumPy namespace and passes u and v as arguments.
    u.dot(v): This is a method call. Here, u is an array object, and dot() is a method defined for NumPy arrays. 
    When you call u.dot(v), it's equivalent to calling np.dot(u, v) but more concise and directly called from the array object u.
"""

u = np.array([1, 2, 3])
v = np.array([1, 10, 100])

print(np.dot(u, v))

# Can also call numpy operations on the numpy array, useful for chaining together multiple operations
print(u.dot(v))

321
321


In [35]:
"""
The code initializes arrays `v` and `W` using NumPy. `v` is a 1D array with 3 elements, and `W` is a 2D array with 3 rows and 2 columns. 
`print(v.shape)` outputs `(3,)`, indicating `v` has 3 elements. `print(W.shape)` outputs `(3, 2)`, showing `W` has 3 rows and 2 columns. 
`np.dot(v, W)` performs matrix multiplication between `v` and `W`, producing a 1D array. 
`print(np.dot(v, W).shape)` outputs `(2,)`, showing the resulting array has 2 elements.
"""

v = np.array([1, 10, 100])
W = np.array([[1, 2], [3, 4], [5, 6]])
print(v.shape)
print(W.shape)

# This works.
print(np.dot(v, W))
print(np.dot(v, W).shape)

(3,)
(3, 2)
[531 642]
(2,)


In NumPy, when performing a dot product between a 1D array (shape: (3,)) and a 2D array (shape: (3, 2)), broadcasting rules are applied. Each element of the 1D array is multiplied element-wise with the corresponding column of the 2D array, and then the products are summed up. The resulting array has the shape of the second operand's dimensions, hence (3, 2) becomes (2,) after the dot product.

For example:

    Element-wise multiplication:
        1 * [1, 2] = [1, 2]
        10 * [3, 4] = [30, 40]
        100 * [5, 6] = [500, 600]
    Summing up the results:
        [1 + 30 + 500, 2 + 40 + 600] = [531, 642]

Thus, the dot product between (3,) and (3, 2) results in [531, 642].

In [16]:
# This does not. Why?
print(np.dot(W, v))

ValueError: shapes (3,2) and (3,) not aligned: 2 (dim 1) != 3 (dim 0)

In [36]:
# We can fix the above issue by transposing W.
print(np.dot(W.T, v))
print(np.dot(W.T, v).shape)

[531 642]
(2,)


###  Indexing

Slicing / indexing numpy arrays is a extension of the Python concept of slicing (lists) to N dimensions.

In [39]:
x = np.random.random((3, 4))

# Selects all of x
print(x[:])
# or print(x)

[[0.16414633 0.60366612 0.70750875 0.151979  ]
 [0.4691696  0.12578704 0.79789923 0.59273757]
 [0.84907456 0.86628498 0.18734879 0.09074287]]


In [18]:
# Selects the 0th and 2nd rows
print(x[np.array([0, 2]), :])

print("---")

# Selects 1st row as 1-D vector and and 1st through 2nd elements
print(x[1, 1:3])

[[0.79576974 0.98783177 0.49225667 0.50055511]
 [0.09517821 0.8439365  0.71366962 0.64966847]]
---
[0.27882605 0.54107395]


In [19]:
# Boolean indexing
print(x[x > 0.5])

[0.79576974 0.98783177 0.50055511 0.54107395 0.8439365  0.71366962
 0.64966847]


In [20]:
# 3-D vector of shape (3, 4, 1)
print(x[:, :, np.newaxis])

[[[0.79576974]
  [0.98783177]
  [0.49225667]
  [0.50055511]]

 [[0.47491422]
  [0.27882605]
  [0.54107395]
  [0.32078553]]

 [[0.09517821]
  [0.8439365 ]
  [0.71366962]
  [0.64966847]]]


### Broadcasting

The term broadcasting describes how NumPy treats arrays with different shapes during arithmetic operations.

**General Broadcasting Rules**

When operating on two arrays, NumPy compares their shapes element-wise. It starts with the trailing (i.e. rightmost) dimensions and works its way left. Two dimensions are compatible when:
- they are equal, or
- one of them is 1 (in which case, elements on the axis are repeated along the dimension)

More details: https://numpy.org/doc/stable/user/basics.broadcasting.html

In [21]:
x = np.random.random((3, 4))

y = np.random.random((3, 1))
z = np.random.random((1, 4))

# In this example, y and z are broadcasted to match the shape of x.
# y is broadcasted along dim 1.
s = x + y
# z is broadcasted along dim 0.
p = x * z

In [22]:
print(x.shape)
print()
print(y.shape)
print(s.shape)

(3, 4)

(3, 1)
(3, 4)


In [23]:
print(x.shape)
print()
print(s.shape)
print(p.shape)

(3, 4)

(3, 4)
(3, 4)


In [24]:
a = np.zeros((3, 3))
b = np.array([[1, 2, 3]])
print(a)
print()
print(a+b)

[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]

[[1. 2. 3.]
 [1. 2. 3.]
 [1. 2. 3.]]


Let's look at a more complex example.

In [25]:
a = np.random.random((3, 4))
b = np.random.random((3, 1))
c = np.random.random((3, ))

What is the expected broadcasting behavior for these operations? What do the following operations give us? What are the resulting shapes?

In [26]:
result1 = b + b.T

print(b.shape)
print(b.T.shape)
print(result1.shape)
print(result1)

(3, 1)
(1, 3)
(3, 3)
[[0.89596095 0.912283   1.10037498]
 [0.912283   0.92860506 1.11669703]
 [1.10037498 1.11669703 1.30478901]]


In [27]:
result2 = a + c

print(a.shape)
print(c.shape)
print(result2.shape)
print(result2)

ValueError: operands could not be broadcast together with shapes (3,4) (3,) 

In [28]:
result3 = b + c

print(b.shape)
print(c.shape)
print(result3.shape)
print(result3)

(3, 1)
(3,)
(3, 3)
[[0.90367481 1.27292439 1.06513149]
 [0.91999687 1.28924645 1.08145355]
 [1.10808884 1.47733842 1.26954552]]


### Efficient NumPy Code

When working with numpy arrays, avoid explicit for-loops over indices/axes at all costs. For-loops will dramatically slow down your code (~10-100x).

We can time code using the %%timeit magic. Let's compare using explicit for-loop vs. using numpy operations.

In [29]:
%%timeit
x = np.random.rand(1000, 1000)
for i in range(100, 1000):
    for j in range(x.shape[1]):
        x[i, j] += 5

411 ms ± 2.14 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [30]:
%%timeit
x = np.random.rand(1000, 1000)
x[np.arange(100,1000), :] += 5

8.96 ms ± 811 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
